fedimint-core/src/db/mod.rs

```diff
@@ -1841,28 +1861,127 @@ pub async fn apply_migrations(
+                } else {
+                    migration(&mut global_dbtx.to_ref_nc()).await?;
+                }
             } else {
                 panic!("Missing migration for version {current_db_version}");
             }
 
             current_db_version.increment();
-            dbtx.insert_entry(&DatabaseVersionKey, &current_db_version)
+            global_dbtx
+                .insert_entry(
+                    &DatabaseVersionKey(module_instance_id_key),
+                    &current_db_version,
+                )
                 .await;
         }
 
         current_db_version
     } else {
-        dbtx.insert_entry(&DatabaseVersionKey, &target_db_version)
-            .await;
         target_db_version
     };
 
-    dbtx.commit_tx_result().await?;
+    global_dbtx.commit_tx_result().await?;
     info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
     Ok(())
 }
 
+/// Migrates the `DatabaseVersion` from inside the module's isolated database
+/// to the global namespace. Database migrations are driven from outside of the
+/// modules so the `DatabaseVersion` should also exist outside the modules.
+/// This also fixes the key prefix conflicts with 0x50.
+pub async fn migrate_database_version(
+    db: &Database,
+    target_db_version: DatabaseVersion,
+    module_instance_id: Option<ModuleInstanceId>,
+    kind: String,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+    let mut global_dbtx = db.begin_transaction().await;
+
+    // Use 0xff for fedimint-server and the `module_instance_id` for each module
+    let key_module_instance_id = if let Some(module_instance_id) = module_instance_id {
+        module_instance_id
+    } else {
+        MODULE_GLOBAL_PREFIX.into()
+    };
+
+    // First check if the module has a `DatabaseVersion` written to
+    // `DatabaseVersionKey` If `DatabaseVersion` already exists, there is
+    // nothing to do.
+    if global_dbtx
+        .get_value(&DatabaseVersionKey(key_module_instance_id))
+        .await
+        .is_none()
+    {
+        // Read and remove the previous `DatabaseVersion`, which is in the module's
+        // isolated namespace (but not for fedimint-server)
+        let current_version_in_module = if let Some(module_instance_id) = module_instance_id {
+            remove_current_db_version_if_exists(
+                &mut global_dbtx
+                    .to_ref_with_prefix_module_id(module_instance_id)
+                    .into_nc(),
+                target_db_version,
+            )
+            .await
+        } else {
+            remove_current_db_version_if_exists(
+                &mut global_dbtx.to_ref().into_nc(),
+                target_db_version,
+            )
+            .await
+        };
+
+        // Write the previous `DatabaseVersion` to the new `DatabaseVersionKey`
+        info!(target: LOG_DB, "Migrating DatabaseVersionKeyV0 of {kind} to DatabaseVersionKey. Version: {current_version_in_module}");
+        global_dbtx
```

Comment:
We should make sure to write db versions for each module on the first run (set to `0`, even if no migrations for them is defined), to make sure this insert happens, and this migration no longer can happen, freeing the key for the module from the start.

Reply:
Yes I believe this is already the case, even if no migration is defined. `migrate_database_version` only does anything interesting if there is no `DatabaseVersionKey` in the database, otherwise it is a no-op. If the old `DatabaseVersionKeyV0` also does not exist, it simply writes `target_db_version` into the database, which will prevent this migration from happening the next time the client is constructed.

Reply:
Only difference from what you said is that I think it needs to write `target_db_version` instead of `0` since that is the current version of the structs that the client is aware of.